scratchml/models/svc.py

```diff
@@ -0,0 +1,202 @@
+            Training data.
+        y : ndarray, shape (n_samples,)
+            Target labels (e.g., 0 or 1 for binary classification).
+        """
+        self.classes_ = np.unique(y)
+        if len(self.classes_) == 2:
+            y_ = np.where(y == self.classes_[0], -1, 1)
+        else:
+            y_ = y
+
+        X = self.scaler.fit_transform(X)
+        n_samples, n_features = X.shape
+        self.weights = np.zeros(n_features)
+
+        no_improvement_count = 0
+        for iteration in range(self.max_iter):
+            lr = self.learning_rate * (self.decay**iteration)
+            avg_update_norm = 0
+            for X_batch, y_batch in split_data_into_batches(
+                X, y_, self.batch_size, shuffle=True
+            ):
+                weight_update = np.zeros_like(self.weights)
+                bias_update = 0
+                for idx in range(X_batch.shape[0]):
+                    instance = X_batch[idx : idx + 1]
+                    margin = y_batch[idx] * (
+                        np.dot(instance.flatten(), self.weights) - self.bias
+                    )
+
+                    if margin < 1:
+                        weight_update += -self.C * y_batch[idx] * instance.flatten()
+                        bias_update += -self.C * y_batch[idx]
+
+                weight_update += self.alpha * l2(self.weights)
+
+                lr_adjusted = (
+                    lr / (1 + np.linalg.norm(weight_update)) if self.adaptive_lr else lr
+                )
+                self.weights -= lr_adjusted * weight_update
+                self.bias -= lr_adjusted * bias_update
+                avg_update_norm += np.linalg.norm(weight_update)
+
+            avg_update_norm /= X.shape[0] / self.batch_size
+            if self.early_stopping and avg_update_norm < self.tol:
+                no_improvement_count += 1
+                if (
+                    no_improvement_count > 10
+                ):  # Terminate if no improvement for 10 iterations
+                    # print(f"Converged after {iteration} iterations.")
+                    break
+            else:
+                no_improvement_count = 0
+
+        #     if iteration % 100 == 0:
+        #         print(f"Iteration {iteration}: Avg norm of batch updates = {avg_update_norm}")
+
+        # print(f"Final weight norm: {np.linalg.norm(self.weights)}")
+        # print(f"Final bias: {self.bias}")
+
+    def predict(self, X):
+        """
+        Predicts class labels for input data.
+        """
+        if self.weights is None:
+            raise ValueError("Model must be trained before prediction.")
+        X = self.scaler.transform(X)
+        linear_output = np.dot(X, self.weights) - self.bias
+        predictions = np.sign(linear_output)
+
+        if len(self.classes_) == 2:
+            return np.where(predictions == -1, self.classes_[0], self.classes_[1])
+        else:
+            return predictions
+
+    def evaluate(self, X, y):
```

Comment:
Please, delete this function, as you already have the `score` function that does the same thing and there's no function `evaluate` in the Scikit-learn's SVC implementation.